Implement `awaitOrTimeout`

mihails-strasuns-sociomantic opened the pull request:
Assuming I haven't missed any dangerous corner case, this should provide the base
of requested implementation.

nemanja-boric-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -124,6 +124,113 @@ unittest
+    theScheduler.eventLoop();
+
+    test(!root.used());
+    test(!root.to_wait_for.used());
```
Any way to write a quick test with waiting for the task that actually finishes?

nemanja-boric-sociomantic replied:
(with no timeout)

mihails-strasuns-sociomantic replied:
Sure, will do (it would be a copy of one for `await` pretty much)

mihails-strasuns-sociomantic replied:
Good catch, I actually had one bug there!
Updated.

nemanja-boric-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -124,6 +124,151 @@ unittest
+    context.suspend();
+    // skip one scheduler event loop cycle to allow awaited task to finish with
+    // all termination hooks
+    theScheduler.processEvents();
```
Haha, yes of course :-)

mihails-strasuns-sociomantic replied:
Actually no, this is not good enough - the same fiber could be busy with other task by that time. I will update to provide required info by task itself.

mihails-strasuns-sociomantic commented:
Updated with what should be more reliable approach to check if task is finished.

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -685,6 +686,9 @@ final class Scheduler
             this.processEvents();
+
+        // prevent referencing old worker fiber
+        task.fiber = null;
```
Is this related to "Introduce task internal state bitmask"?

mihails-strasuns-sociomantic replied:
Yes, though a bit indirectly. `task.fiber is null` with thi can be used to check for extra state - "not used at all".

gavin-norman-sociomantic commented on src/ocean/task/Task.d:
```diff
@@ -383,27 +384,42 @@ public abstract class Task : ISuspendable
+
+    ***************************************************************************/
+
+    final public bool finished ( )
```
Release note: new feature.

gavin-norman-sociomantic commented on src/ocean/task/Task.d:
```diff
@@ -452,6 +471,7 @@ public abstract class Task : ISuspendable
         catch (TaskKillException)
         {
             debug_trace("<{}> termination (killed)", cast(void*) this);
+            this.state_bitmask &= ~TaskState.ToKill;
```
Interesting that `to_kill` was never reset anywhere before...!

mihails-strasuns-sociomantic replied:
Don't bring attention to my failures gone unnoticed :P

gavin-norman-sociomantic commented on relnotes/await-timeout.feature.md:
```diff
@@ -0,0 +1,16 @@
+
+  New function `awaitOrTimeout` can be used to suspend some task until either
+  some other specified task finished OR until timer is hit, whichever comes
+  first. Conceptually is is similar to `theScheduler.await`, but with a timeout.
```
Typo "it is"

gavin-norman-sociomantic commented on relnotes/await-timeout.feature.md:
````diff
@@ -0,0 +1,16 @@
+  ```D
+  auto task = new SomeTask;
+  bool timeout = .awaitOrTimeout(task, 1_000_000);
+  // Doesn't kill on timeout by default, as it is not needed with tasks like DHT
````
Though we're not prohibited from mentioning things like the DHT / swarm in public now, it feels weird to refer to it in ocean, which is a lower level library.

gavin-norman-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -124,6 +124,149 @@ unittest
+    happens first.
+
+    Params:
+        task = task to await for
```
Grammar tip of the day: "await" or "wait for" (not "await for").

gavin-norman-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -124,6 +124,149 @@ unittest
+{
+    initScheduler(SchedulerConfiguration.init);
+
+    .timer = new typeof(timer);
```
Hmm seems unnecessary.

gavin-norman-sociomantic replied:
Unnecessary?

mihails-strasuns-sociomantic replied:
Necessay until timer setup code gets rewritten in next commit. Depending on order of test block being run, `.timer` could be null here.

gavin-norman-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -83,7 +80,27 @@ unittest
     initScheduler(SchedulerConfiguration.init);
 
-    .timer = new typeof(timer);
+    class SimpleTask : Task
```
Is this related to "Create global timer lazily inside helper"?

mihails-strasuns-sociomantic replied:
Yes, diff is confusing. I have split unittest that requires resetting timer instance manually to check allocated slot count from documented unittest which uses regular lazily created timer.

gavin-norman-sociomantic reviewed (commented):
Nice!

mihails-strasuns-sociomantic commented:
Updated with doc and release note fixes

mihails-strasuns-sociomantic commented:
Ping @gavin-norman-sociomantic

mihails-strasuns-sociomantic commented:
Ping @gavin-norman-sociomantic

gavin-norman-sociomantic reviewed (commented):
Just the question about initialising the timer in unittests.

gavin-norman-sociomantic commented on src/ocean/task/util/Timer.d:
```diff
@@ -83,7 +80,27 @@ unittest
+
+    initScheduler(SchedulerConfiguration.init);
+
+    .timer = new typeof(.timer);
```
And is it necessary here?

mihails-strasuns-sociomantic replied:
Yes, for a reason explained above (it tests exact allocated slot count)

gavin-norman-sociomantic replied:
Hmm, I don't understand that comment. What's a "slot"? Why is it necessary to re-allocate the timer? Doc enhancement please :)

mihails-strasuns-sociomantic replied:
https://github.com/sociomantic-tsunami/ocean/pull/234/files#diff-a3ce7f2ae79a2097ba597486e0d9a30fR121 + comment 3 lines above

Which doc?

gavin-norman-sociomantic replied:
Some documentation of what this unittest is testing. It's not clear to me.

>Why is it necessary to re-allocate the timer?

"Reallocate a fresh timer so the test result is not affected by previous unittests" ?

mihails-strasuns-sociomantic replied:
Updated

mihails-strasuns-sociomantic committed:
Introduce task internal state bitmask

Replaces plain `bool to_kill` as there are now new kinds of flags to
keep and task object must be as lightweight as possible.

mihails-strasuns-sociomantic committed:
Implement `awaitOrTimeout`

Fixes #46

mihails-strasuns-sociomantic committed:
Refactor common timer code into private function

mihails-strasuns-sociomantic committed:
Create global timer lazily inside helper

Now that there is only one function that will ever access it directly in
non-test code, better to delegate all interaction with timer to it.

mihails-strasuns-sociomantic committed:
Fix flakey `FileException` doc test

Using hard-coded 42 as invalid file descriptor is wrong as such
descriptor number can totally exist in the test suite.
